Issue #339. with tn.DefaultBackend(backend): support

katolikyan opened the pull request:
I tried to implement this one. I believe the DefaultBackend class could be better but still :)

I also have tried to implement it in another way, more like a temporary change for `config.default_backend` in a `with` statement. In this way, it doesn't affect any code inside the `class Node`, but if someone will set `config.default_backend` inside a `with` statement that will cause a little mess. Here is the main idea how would another approach look like:

```python3
from typing import Text, Union
from contextlib import contextmanager
from tensornetwork.backends.base_backend import BaseBackend
import tensornetwork.config as config
from tensornetwork.__init__ import set_default_backend

@contextmanager
def DefaultBackend(backend: Union[Text, BaseBackend]) -> None:
    try:
        default_backend = config.default_backend
        set_default_backend(backend)
        yield None
    finally:
        set_default_backend(default_backend)
```

katolikyan commented:
@googlebot I signed it!

katolikyan commented:
@googlebot I fixed it.

Thenerdstation commented on tensornetwork/backend_contextmanager.py:
```diff
@@ -0,0 +1,17 @@
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        DefaultBackend.with_statement = False
+        DefaultBackend.backend = None
```
Nice! I like the object oriented approach.

However, I think we should be using a stack based approach for finding the current backend. (For an example, see what we did with the `NodeCollection` [code](https://github.com/google/TensorNetwork/blob/83ff4af11dcd03d6c0db70e139c7137154dc015d/tensornetwork/network_components.py#L1978).

Thenerdstation commented on tensornetwork/tests/backend_contextmanager_test.py:
```diff
@@ -0,0 +1,32 @@
+        b = tn.Node(np.ones((10,)))
+
+    assert (type(a.backend) == type(b.backend)) == True
+    assert tn.config.default_backend == "tensorflow"
```
We should also consider the case where we have nested `with` statements. You can imagine this happening for when a library needs to implement something purely in numpy or when a user wants to test their code in all backends.

Maybe add a test like this?

```python
with tn.DefaultBackend("tensorflow"):
  a = tn.Node(...)
  assert a.backend.name == "tensorflow" 
  with tn.DefaultBackend("numpy"):
     b = tn.Node(...)
     assert b.backend.name == "numpy"
  c = tn.Node(...)
  assert c.backend.name == "tensorflow"
```

katolikyan commented:
Thank you for the comments! I tried to implement it with a stack as you showed in the example.
I also have added a couple of tests.
But I am not sure if I should have written the `DefaultBackend` class inside the `network_components.py` and put the stack into the `ops.py` file.

Thenerdstation commented on tensornetwork/tests/backend_contextmanager_test.py:
```diff
@@ -0,0 +1,53 @@
+  with tn.DefaultBackend("numpy"):
+    assert tn.config.default_backend == "pytorch"
+    a = tn.Node(np.ones((10,)))
+    tn.set_default_backend("tensorflow")
```
using `set_default_backend` while there is a backend stack feels like a foot gun. Can you raise an error in this case instead?

katolikyan replied:
Sorry I didn't get what you mean by raising an error in this case. Prohibit using `set_default_backend` while in `with` statement?

Thenerdstation commented on tensornetwork/network_components.py:
```diff
@@ -516,7 +517,11 @@ def __init__(self,
+      if default_backend:
+        backend = default_backend.backend
+      else:
+        backend = config.default_backend
```
Can this logic be moved to the `DefaultBackend` code?

Thenerdstation commented on tensornetwork/backend_contextmanager.py:
```diff
@@ -0,0 +1,28 @@
+    self.stack = []
+
+  def get_current_backend(self):
+    return self.stack[-1] if self.stack else None
```
Move the logic in network_components to here.

katolikyan commented:
I moved the logic to the right place. and cleaned up a test a little to get rid of some unnecessary checks.

Thenerdstation commented on tensornetwork/tests/backend_contextmanager_test.py:
```diff
@@ -0,0 +1,42 @@
+def test_contextmanager_wrong_item():
+  a = tn.Node(np.ones((10,)))
+  try:
+    with tn.DefaultBackend(a):
```
Add `# pytype: disable=wrong-arg-types` to this line to pass the build.

Thenerdstation commented on tensornetwork/tests/backend_contextmanager_test.py:
```diff
@@ -0,0 +1,42 @@
+      pass
+    assert False
+  except ValueError:
+    assert True
```
You should use `with pytest.raises(ValueError)` instead.

https://docs.pytest.org/en/latest/assert.html

Thenerdstation commented on tensornetwork/backend_contextmanager.py:
```diff
@@ -0,0 +1,32 @@
+from typing import Text, Union
+from tensornetwork.backends.base_backend import BaseBackend
+import tensornetwork.config as config
```
Can we move the `set_default_backend` and everything in `config` to this file? No need to make these separate anymore and I like this filename more.

Thenerdstation commented on tensornetwork/__init__.py:
```diff
@@ -15,6 +15,7 @@
+from tensornetwork.backend_contextmanager import DefaultBackend
 
 
 def set_default_backend(backend: Union[Text, BaseBackend]) -> None:
```
Add a check that if the default backend stack is empty, and if it's not, raise an `AssertionError`. 

This is to help avoid the confusing situation like: 

```python
with tn.DefaultBackend("tensorflow"):
  tn.set_default_backend("jax")
  a = tn.Node(...)
```
Should `a` be using jax or tensorflow? It's confusing and kinda a footgun, so let's just crash with a clear message instead.

katolikyan commented:
Hopefully, I haven't messed up the code. Removed `config` imports with the thought that it became useless in this case. But feel like maybe I shouldn't have done that.

Also, from my understanding, we don't need direct access to `default_backend` anymore, do we?
I moved it to the `_DefaultBackendStack` class, trying to avoid using `global`. I can quickly rewrite it as a global with something like this if needed :)

```python3
default_backend = "numpy"

def set_default_backend(backend: Union[Text, BaseBackend]) -> None:
  global default_backend
  if _default_backend_stack.stack:
    raise AssertionError("The default backend should not be changed "
                         "inside the backend context manager")
  if not isinstance(backend, (Text, BaseBackend)):
    raise ValueError("Item passed to set_default_backend "
                     "must be Text or BaseBackend")
  default_backend = backend
```

Thenerdstation commented:
Looks awesome! Thank you so much! I've been waiting on this one for a while. ^-^

katolikyan commented:
Glad to hear this :) it was nice to work on that issue, thank you! :)
